### PR TITLE
Create action added for Parking Slot

### PR DESCRIPTION
**Pull request by abdullokhtolibjonov**

### CODE REVIEW

File: Parking-Zone/Areas/Admin/Views/ParkingSlot/Create.cshtml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+<h1>Create</h1>
+
+<h4>ParkingSlot</h4>
+<hr />
+<div class="row">
+    <div class="col-md-4">
+        <form asp-action="Create">
+            <div asp-validation-summary="ModelOnly" class="text-danger"></div>
+            <input type="hidden" value="@Model.ParkingZoneId" asp-for="ParkingZoneId" />
+            <input type="hidden" value="@Model.ParkingZoneName" asp-for="ParkingZoneName" />
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Views/ParkingSlot/Create.cshtml]: You don't need this property to be submitted.

File: Parking-Zone/Services/ParkingSlotService.cs [hunk 2]
```diff
@@ -15,5 +15,16 @@ public IEnumerable<ParkingSlot> GetByParkingZoneId(Guid parkingZoneId)
...
+        public override void Insert(ParkingSlot slot)
+        {
+            slot.Id = Guid.NewGuid();   
+            base.Insert(slot);
+        }
+
+        public bool SlotExistsWithThisNumber(int slotNumber, Guid? slotId, Guid parkingZoneId)
+            => _parkingSlotRepository.GetAll()
+            .Where(s => s.ParkingZoneId == parkingZoneId && s.Id != slotId)
+            .FirstOrDefault(s => s.Number == slotNumber) != null;
```
Comment (napster1991) [inline Parking-Zone/Services/ParkingSlotService.cs]:
```suggestion
                .Where(s => s.ParkingZoneId == parkingZoneId && s.Id != slotId)
                .FirstOrDefault(s => s.Number == slotNumber) != null;
```

File: Parking-Zone/ViewModels/ParkingSlot/ParkingSlotCreateVM.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+{
+    public class ParkingSlotCreateVM
+    {
+        [Required]
+        public int Number { get; set; }
+        [Required]
+        public bool IsAvailableForBooking { get; set; }
+        [Required]
+        public SlotCategoryEnum Category { get; set; }
+        public Guid ParkingZoneId { get; set; }
```
Comment (napster1991) [inline Parking-Zone/ViewModels/ParkingSlot/ParkingSlotCreateVM.cs:15]: Also required

File: Tests/Controllers/ParkingSlotControllerTests.cs [hunk 4]
```diff
@@ -108,7 +110,93 @@ public void GivenNullParkingZoneId_WhenIndexIsCalled_ThenParkingZoneServiceIsCal
...
+            mockZoneService.Verify(s => s.GetById(Guid.Parse("")), Times.Once);
+        }
+        #endregion
+
+        #region Create
+        [Fact]
+        public void GivenParkingZoneId_WhenGetCreateIsCalled_ThenParkingZoneServiceIsCalledAndReturnedNotEmptyViewResult()
+        {
+            //Arrange
+            mockZoneService.Setup(service => service.GetById(_testZoneId)).Returns(_testZone);
```
Comment (napster1991) [inline Tests/Controllers/ParkingSlotControllerTests.cs]:
```suggestion
            mockZoneService
                .Setup(service => service.GetById(_testZoneId))
                .Returns(_testZone);
```

File: Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 5]
```diff
@@ -8,32 +8,74 @@ namespace Parking_Zone.Areas.Admin.Controllers
...
+            ViewData["parkingZoneName"] = Zone.Name;
+            ViewData["parkingZoneId"] = zoneId;
 
             return View(slotVMs);
         }
+
+        // GET: Admin/ParkingSlots/Create
+        public IActionResult Create(Guid zoneId)
+        {
+            var Zone = _zoneService.GetById(zoneId);
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs]:
```suggestion
            var zone = _zoneService.GetById(zoneId);
```

File: Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 6]
```diff
@@ -8,32 +8,74 @@ namespace Parking_Zone.Areas.Admin.Controllers
...
+
+            if (Zone is null)
+            {
+                return BadRequest();
+            }
+
+            var SlotCreateVM = new ParkingSlotCreateVM();
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs]:
```suggestion
            var slotCreateVM = new ParkingSlotCreateVM();
```

File: Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 7]
```diff
@@ -8,32 +8,74 @@ namespace Parking_Zone.Areas.Admin.Controllers
...
+        public IActionResult Create(ParkingSlotCreateVM slotCreateVM)
+        {
+            if (_slotService.SlotExistsWithThisNumber(slotCreateVM.Number, null, slotCreateVM.ParkingZoneId))
+            {
+                ModelState.AddModelError("Number", "This Slot number already exists");
+            }
+
+            if (ModelState.IsValid)
+            {
+                var Zone = _zoneService.GetById(slotCreateVM.ParkingZoneId);
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs]:
```suggestion
                var zone = _zoneService.GetById(slotCreateVM.ParkingZoneId);
```

File: Tests/Controllers/ParkingSlotControllerTests.cs [hunk 8]
```diff
@@ -108,7 +110,93 @@ public void GivenNullParkingZoneId_WhenIndexIsCalled_ThenParkingZoneServiceIsCal
...
+            //Arrange
+            var slotCreateVM = new ParkingSlotCreateVM()
+            {
+                Number = 1,
+                ParkingZoneId = _testZoneId,
+                Category = SlotCategoryEnum.Business,
+                IsAvailableForBooking = true
+            };
+
+            mockSlotService.Setup(service => service.Insert(It.IsAny<ParkingSlot>()));
```
Comment (napster1991) [inline Tests/Controllers/ParkingSlotControllerTests.cs]:
```suggestion
            mockSlotService
                .Setup(service => service.Insert(It.IsAny<ParkingSlot>()));
```

File: Tests/Controllers/ParkingSlotControllerTests.cs [hunk 9]
```diff
@@ -108,7 +110,93 @@ public void GivenNullParkingZoneId_WhenIndexIsCalled_ThenParkingZoneServiceIsCal
...
+
+            //Act
+            var result = controller.Create(_testZoneId);
+
+            //Assert
+            Assert.IsType<ViewResult>(result);
+            Assert.NotNull((result as ViewResult).Model);
+            mockZoneService.Verify(s => s.GetById(_testZoneId), Times.Once);
+        }
+
```
Comment (napster1991) [inline Tests/Controllers/ParkingSlotControllerTests.cs:142]: Add test case for action GET: Create when zoneId parameter is null

File: Tests/Controllers/ParkingSlotControllerTests.cs [hunk 10]
```diff
@@ -108,7 +110,93 @@ public void GivenNullParkingZoneId_WhenIndexIsCalled_ThenParkingZoneServiceIsCal
...
+            //Act
+            var result = controller.Create(slotCreateVM);
+
+            //Assert
+            Assert.False(controller.ModelState.IsValid);
+            Assert.IsType<ViewResult>(result);
+            Assert.Equal(JsonSerializer.Serialize(slotCreateVM), JsonSerializer.Serialize((result as ViewResult).Model));
+        }
+
```
Comment (napster1991) [inline Tests/Controllers/ParkingSlotControllerTests.cs:209]: Add test cases:
1. GivenModelWithAlreadyExistingNumber_***
2. GivenModelWithNullZoneId_WhenPostCreateIsCalled_ThenCreateViewReturnedAndModelStateIsInvalid()

File: Tests/Services/ParkingSlotServiceTests.cs [hunk 11]
```diff
@@ -50,20 +50,37 @@ public void GivenParkingZoneId_WhenGetByParkingZoneId_ThenRepositoryIsCalledTwic
...
-            var result = service.GetByParkingZoneId(_testParkingZoneId);
+            var result = service.GetByParkingZoneId(_testZoneId);
 
             //Assert
             Assert.Equal(2, result.Count());
-            mockParkingSlotRepository.Verify(repo => repo.GetAll(), Times.Once);
+            mockSlotRepository.Verify(repo => repo.GetAll(), Times.Once);
+        }
+        #endregion
+
```
Comment (napster1991) [inline Tests/Services/ParkingSlotServiceTests.cs:69]: Where are tests for SlotExistsWithThisNumber method?

**Review by napster1991 — CHANGES_REQUESTED** [review]

File: Tests/Controllers/ParkingSlotControllerTests.cs [hunk 12]
```diff
@@ -108,7 +110,93 @@ public void GivenNullParkingZoneId_WhenIndexIsCalled_ThenParkingZoneServiceIsCal
...
+            mockZoneService.Setup(service => service.GetById(_testZoneId)).Returns(_testZone);
+
+            //Act
+            var result = controller.Create(_testZoneId);
+
+            //Assert
+            Assert.IsType<ViewResult>(result);
+            Assert.NotNull((result as ViewResult).Model);
+            mockZoneService.Verify(s => s.GetById(_testZoneId), Times.Once);
+        }
```
Comment (napster1991) [inline Tests/Controllers/ParkingSlotControllerTests.cs:141]: Add such assertion: `Assert.Equal(JsonSerializer.Serialize(expectedCreateSlotVM), JsonSerializer.Serialize((result as ViewResult).Model));`. See Dilshod's [PR](https://github.com/DilshodMQ/Parking-zone/pull/9/files#diff-0be201021cb04ff536474e37b39372b678b1d2ab895950064dab90d36337ae85R142) for details

File: Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 13]
```diff
@@ -8,32 +8,74 @@ namespace Parking_Zone.Areas.Admin.Controllers
...
+
+                if (Zone is null)
+                    return BadRequest("Parking Zone Not Found");
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs]: Add test for this case. See [here](https://github.com/DilshodMQ/Parking-zone/pull/9/files#diff-0be201021cb04ff536474e37b39372b678b1d2ab895950064dab90d36337ae85R228)

**Review by napster1991 — CHANGES_REQUESTED** [review]

**Review by napster1991 — APPROVED** [review]
